src/backend/distributed/planner/distributed_planner.c

```diff
@@ -118,7 +122,25 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 	}
 	else if (CitusHasBeenLoaded())
 	{
-		needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList);
+		bool hasReferenceTableReplica = false;
+
+		/*
+		 * All groups that have pg_dist_node entries, also have reference
+		 * table replicas.
+		 */
+		PrimaryNodeForGroup(GetLocalGroupId(), &hasReferenceTableReplica);
+
+		if (IsCoordinator() &&
+			hasReferenceTableReplica &&
+			IsLocalReferenceTableJoin(parse, rangeTableList))
```

Comment:
Maybe we can put these checks in a function?

Reply:
done